vdf/src/create_discriminant.rs

```diff
@@ -58,26 +55,31 @@ fn random_bytes_from_seed(seed: &[u8], byte_count: usize) -> Vec<u8> {
 /// This function is guaranteed not to panic for any inputs whatsoever, unless
 /// memory allocation fails and the allocator in use panics in that case.
 pub fn create_discriminant(seed: &[u8], length: u16) -> Mpz {
-    let extra: u8 = (length as u8) & 7;
-    let random_bytes_length = ((length >> 3) + if extra == 0 { 2 } else { 3 }) as usize;
-    let mut random_bytes = random_bytes_from_seed(seed, random_bytes_length);
-    assert_eq!(random_bytes.len(), random_bytes_length);
-    let residue = {
-        let last_2 = &mut random_bytes[random_bytes_length - 2..];
+    let (mut n, residue) = {
+        // The number of “extra” bits (that don’t evenly fit in a byte)
+        let extra: u8 = (length as u8) & 7;
+
+        // The number of random bytes needed (the number of bytes that hold `length` bits,
+        // plus 2).
+        let random_bytes_len = ((usize::from(length) + 7) >> 3) + 2;
+        let random_bytes = random_bytes_from_seed(seed, random_bytes_len);
+        let (n, last_2) = random_bytes.split_at(random_bytes_len - 2);
         let numerator = (usize::from(last_2[0]) << 8) + usize::from(last_2[1]);
-        RESIDUES[numerator % RESIDUES.len()]
+
+        // If there are any extra bits, right shift `n` so that it fits
+        // in `length` bits, discarding the least significant bits.
+        let n = Mpz::from(n) >> usize::from((8 - extra) & 7);
+        (n, RESIDUES[numerator % RESIDUES.len()])
     };
-    let orig_n = Mpz::from(&random_bytes[..random_bytes_length as usize - 2usize]);
-    drop(random_bytes);
-    let mut n: Mpz = &orig_n >> ((8 - extra) & 7).into();
-    n.setbit((length - 1) as _);
-    {
-        let q: Mpz = M.into();
-        let r: Mpz = &n % q;
-        n -= r;
-    }
-    {
-        mpz_add_ui_self(&mut n, residue);
+    n.setbit(usize::from(length - 1));
+    n = -n;
+    let rem = mpz_rem_u32(&n, M);
+    n = -n;
```

Comment:
Is there a reason not just to

```rust
let rem = mpz_rem_u32(&(-n), M);
```

and avoid reassigning `n` twice?

Reply:
I suspect that doing it this way avoids a copy.

Reply:
Also, `mpz_rem_u32` is a (very) poor name ― it calls `__gmpz_cdiv_ui`, rather than `__gmpz_tdiv_ui` as would be expected.

Reply:
To elaborate: GMP uses sign-magnitude internally, so flipping the sign of a number is cheap.  Furthermore, `n = -n` actually modifies `n` , rather than needing to copy it.